signingscript/src/signingscript/sign.py

```diff
@@ -1548,3 +1548,38 @@ async def sign_debian_pkg(context, path, fmt, *args, **kwargs):
             output_file.write(signed_file["content"])
     await _create_tarfile(context, path, all_file_names, compression, tmp_dir=tmp_dir)
     return path
+
+
+async def rcodesign_notarize(paths, notarization_workdir, creds_path):
+    NOTARIZATION_EXTENSIONS = ("app", "pkg")
+    for app in paths:
+        if app.split(".")[-1].lower() not in NOTARIZATION_EXTENSIONS:
+            raise SigningScriptError("Unable to notarize app format: %s" % app)
+        app_path = os.path.join(notarization_workdir, app)
+        command = [
+            "rcodesign",
+            "notary-submit",
+            "--staple",
+            "--api-key-path",
+            creds_path,
+            app_path,
+        ]
+        await utils.execute_subprocess(command)
+
+
+@time_async_function
+async def apple_notarize(context, path, *args, **kwargs):
+    """
+    Notarizes given package(s) using rcodesign.
+    """
+    # Setup workdir
+    notarization_workdir = os.path.join(context.config["work_dir"], "apple_notarize")
+    _, compression = os.path.splitext(path)
+    all_file_names = await _extract_tarfile(context, path, compression, tmp_dir=notarization_workdir)
+    workdir_files = os.listdir(notarization_workdir)
+
+    # Notarize
+    await rcodesign_notarize(workdir_files, notarization_workdir, context.apple_credentials_path)
+
+    # Compress files and return path to tarball
+    return await _create_tarfile(context, path, all_file_names, compression, notarization_workdir)
```

Comment:
Is there a particular reason we're sending these back as tarballs instead of pkg or apps (as we do with authenticode signing)?

Reply:
`.app`s are not compressed at all.
Since this is a "new" type of task, defaulting to a compressed output seems reasonable.
That being said, we might adjust it later if required 🙃

Reply:
Ah right - I totally lost the plot on `.app` format :laughing: . This is fine.